### PR TITLE
Alerting: Optimize external Loki queries

### PR DESCRIPTION
**Pull request by JohnnyQQQQ**
This PR also enables the optimization of external loki queries. We had a few users running external queries from another stack in cloud, that would not be optimized. So instead of only allowing the provisioned datasource, we also will enable the optimization on any arbitrary loki query.

### CODE REVIEW

File: pkg/services/ngalert/store/loki_range_to_instant.go [hunk 1]
```diff
@@ -20,20 +31,27 @@ func canBeInstant(r *models.AlertRule) bool {
...
+		!t.isLoki() {
 		return false
 	}
+
 	exprRaw := make(map[string]interface{})
 	if err := json.Unmarshal(r.Data[1].Model, &exprRaw); err != nil {
 		return false
 	}
+	// Second query part should be and expression, '-100' is the legacy way to define it.
+	if r.Data[1].DatasourceUID != "__expr__" && r.Data[1].DatasourceUID != "-100" {
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/store/loki_range_to_instant.go:51]: I appreciated this code was just moved around, but might as well change these to use the function:
```suggestion
	if expr.IsDataSource(r.Data[1].DatasourceUID) {
```

File: pkg/services/ngalert/store/loki_range_to_instant.go [hunk 2]
```diff
@@ -12,6 +12,17 @@ const (
...
 
+// DSType can be used to check the datasource type if it's set in the model.
+type dsType struct {
+	DS struct {
+		Type string `json:"type"`
+	} `json:"datasource"`
+}
+
+func (t dsType) isLoki() bool {
+	return t.DS.Type == "loki"
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/store/loki_range_to_instant.go:23]: Can we use `datasources.DS_LOKI` here?

File: pkg/services/ngalert/store/loki_range_to_instant.go [hunk 3]
```diff
@@ -20,20 +31,27 @@ func canBeInstant(r *models.AlertRule) bool {
...
+
 	if r.Data[0].DatasourceUID != grafanaCloudLogs &&
 		r.Data[0].DatasourceUID != grafanaCloudUsageInsights &&
-		r.Data[0].DatasourceUID != grafanaCloudStateHistory {
-		return false
-	}
-	// Second query part should be and expression, '-100' is the legacy way to define it.
-	if r.Data[1].DatasourceUID != "__expr__" && r.Data[1].DatasourceUID != "-100" {
+		r.Data[0].DatasourceUID != grafanaCloudStateHistory &&
+		!t.isLoki() {
```
Comment (alexweav) [inline pkg/services/ngalert/store/loki_range_to_instant.go:42]: Why keep all the UID checks? They're all already loki, can this be simplified to `if !t.isLoki() { return false }`?

File: pkg/services/ngalert/store/loki_range_to_instant.go [hunk 4]
```diff
@@ -12,6 +12,17 @@ const (
 	grafanaCloudStateHistory  = "grafanacloud-loki-alert-state-history"
 )
```
Comment (alexweav) [inline pkg/services/ngalert/store/loki_range_to_instant.go:13]: Can be removed if we follow my other comment about not checking this

**Review by alexweav — APPROVED** [review]
LGTM once my and George's minor comments are addressed - i don't feel the need to re review

File: pkg/services/ngalert/store/loki_range_to_instant.go (hunk 3 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/store/loki_range_to_instant.go:44]: I thought about it but I was not sure if we always add the type, e.g. during migrations etc. But I think if we three all agree it should be fine. Will change it.

**Review by grobinson-grafana — APPROVED** [review]

**Pull request merged**